### PR TITLE
Mid: fenced: Fencing escalation may not be performed if there is no topology.

### PR DESCRIPTION
**Pull request by HideoYamauchi**
Hi All,

Execution of fencing escalation consisting of N nodes without topology settings may fail.
Occurs when processed by "static-list".
If there is a topology, this problem does not apply because the processing is different.

Here are the successes and failures: The following example specifies pcmk_host_list.
The virtual environment is taken as an example below, but the same is true for the real environment.

Also, depending on the version of fence_agent, disable-timeout may be enabled, so it is necessary to disable disable-timeout for reproduction.

```
(snip)
pcs stonith create fence1-virsh fence_virsh \
    pcmk_reboot_action="reboot" pcmk_host_list="rh83-dev01" ip="192.168.3.253" username="root" password="xxxxx" \
    op start timeout=60s on-fail=restart \
    monitor timeout=60s interval=3600s on-fail=restart \
    stop timeout=60s on-fail=ignore

pcs stonith create fence2-virsh fence_virsh \
    pcmk_reboot_action="reboot" pcmk_host_list="rh83-dev02" ip="192.168.3.253" username="root" password="xxxxx" \
    op start timeout=60s on-fail=restart \
    monitor timeout=60s interval=3600s on-fail=restart \
    stop timeout=60s on-fail=ignore

pcs stonith create fence3-virsh fence_virsh \
    pcmk_reboot_action="reboot" pcmk_host_list="rh83-dev03" ip="192.168.3.253" username="root" password="xxxxx" \
    op start timeout=60s on-fail=restart \
    monitor timeout=60s interval=3600s on-fail=restart \
    stop timeout=60s on-fail=ignore
(snip)
pcs constraint location fence1-virsh avoids rh83-dev01
pcs constraint location fence2-virsh avoids rh83-dev02
pcs constraint location fence3-virsh avoids rh83-dev03
(snip)

```

Successful case.
Step1) Configure the cluster.
```
[root@rh83-dev01 ~]# crm_mon -rfA1
Cluster Summary:
  * Stack: corosync
  * Current DC: rh83-dev02 (version 2.1.0-f7da24cd0) - partition with quorum
  * Last updated: Mon Jun  7 09:02:54 2021
  * Last change:  Mon Jun  7 08:58:49 2021 by root via cibadmin on rh83-dev01
  * 3 nodes configured
  * 7 resource instances configured

Node List:
  * Online: [ rh83-dev01 rh83-dev02 rh83-dev03 ]

Full List of Resources:
  * fence1-virsh        (stonith:fence_virsh):   Started rh83-dev02
  * fence2-virsh        (stonith:fence_virsh):   Started rh83-dev01
  * fence3-virsh        (stonith:fence_virsh):   Started rh83-dev01
  * Resource Group: dummy-group:
    * dummy-1   (ocf:heartbeat:Dummy):   Started rh83-dev01
  * Clone Set: prmPing-clone [prmPing]:
    * Started: [ rh83-dev01 rh83-dev02 rh83-dev03 ]

Node Attributes:
  * Node: rh83-dev01:
    * default_ping_set                  : 1         
  * Node: rh83-dev02:
    * default_ping_set                  : 1         
  * Node: rh83-dev03:
    * default_ping_set                  : 1         

Migration Summary:
```

Step2) Blocks the fencing connection path for node 2, which preferentially fencing node 1.
```
[root@rh83-dev02 ~]# iptables -A INPUT -s 192.168.3.253 -j DROP;iptables -A OUTPUT -s 192.168.3.253 -j DROP
```

Step3) Causes a resource outage failure on node 1.

Step4) Fencing from node 2 will fail, but fencing from node 3 will be performed.

Step5) As a result, it fails over.
```
[root@rh83-dev02 ~]# crm_mon -rfA1
Cluster Summary:
  * Stack: corosync
  * Current DC: rh83-dev02 (version 2.1.0-f7da24cd0) - partition with quorum
  * Last updated: Mon Jun  7 09:05:35 2021
  * Last change:  Mon Jun  7 08:58:49 2021 by root via cibadmin on rh83-dev01
  * 3 nodes configured
  * 7 resource instances configured

Node List:
  * Online: [ rh83-dev02 rh83-dev03 ]
  * OFFLINE: [ rh83-dev01 ]

Full List of Resources:
  * fence1-virsh        (stonith:fence_virsh):   Started rh83-dev02
  * fence2-virsh        (stonith:fence_virsh):   Started rh83-dev03
  * fence3-virsh        (stonith:fence_virsh):   Stopped
  * Resource Group: dummy-group:
    * dummy-1   (ocf:heartbeat:Dummy):   Started rh83-dev02
  * Clone Set: prmPing-clone [prmPing]:
    * Started: [ rh83-dev02 rh83-dev03 ]
    * Stopped: [ rh83-dev01 ]

Node Attributes:
  * Node: rh83-dev02:
    * default_ping_set                  : 1         
  * Node: rh83-dev03:
    * default_ping_set                  : 1         

Migration Summary:
  * Node: rh83-dev02:
    * fence3-virsh: migration-threshold=1 fail-count=1000000 last-failure='Mon Jun  7 09:05:07 2021'

Failed Resource Actions:
  * fence3-virsh_start_0 on rh83-dev02 'error' (1): call=27, status='complete', exitreason='', last-rc-change='2021-06-07 09:04:56 +09:00', queued=0ms, exec=10353ms
```

Case of failure.
Step1) Configure the cluster.
```
[root@rh83-dev01 ~]# crm_mon -rfA1
Cluster Summary:
  * Stack: corosync
  * Current DC: rh83-dev01 (version 2.1.0-f7da24cd0) - partition with quorum
  * Last updated: Mon Jun  7 09:13:51 2021
  * Last change:  Mon Jun  7 09:13:42 2021 by root via cibadmin on rh83-dev01
  * 3 nodes configured
  * 7 resource instances configured

Node List:
  * Online: [ rh83-dev01 rh83-dev02 rh83-dev03 ]

Full List of Resources:
  * fence1-virsh        (stonith:fence_virsh):   Started rh83-dev02
  * fence2-virsh        (stonith:fence_virsh):   Started rh83-dev01
  * fence3-virsh        (stonith:fence_virsh):   Started rh83-dev01
  * Resource Group: dummy-group:
    * dummy-1   (ocf:heartbeat:Dummy):   Started rh83-dev01
  * Clone Set: prmPing-clone [prmPing]:
    * Started: [ rh83-dev01 rh83-dev02 rh83-dev03 ]

Node Attributes:
  * Node: rh83-dev01:
    * default_ping_set                  : 1         
  * Node: rh83-dev02:
    * default_ping_set                  : 1         
  * Node: rh83-dev03:
    * default_ping_set                  : 1         

Migration Summary:
```
Step2) Blocks the fencing connection path for node 2, which preferentially fencing node 1.
```
[root@rh83-dev02 ~]# iptables -A INPUT -s 192.168.3.253 -j DROP;iptables -A OUTPUT -s 192.168.3.253 -j DROP
```
Step3) Causes a resource outage failure on node 1.

Step4) Fencing from node 2 will fail, but fencing from node 3 will not be performed.

Step5) As a result, failover fails.
```

[root@rh83-dev01 ~]# crm_mon -rfA1m3
Cluster Summary:
  * Stack: corosync
  * Current DC: rh83-dev01 (version 2.1.0-f7da24cd0) - partition with quorum
  * Last updated: Mon Jun  7 09:16:20 2021
  * Last change:  Mon Jun  7 09:13:42 2021 by root via cibadmin on rh83-dev01
  * 3 nodes configured
  * 7 resource instances configured

Node List:
  * Node rh83-dev01: UNCLEAN (online)
  * Online: [ rh83-dev02 rh83-dev03 ]

Full List of Resources:
  * fence1-virsh        (stonith:fence_virsh):   Started rh83-dev02
  * fence2-virsh        (stonith:fence_virsh):   Started [ rh83-dev01 rh83-dev03 ]
  * fence3-virsh        (stonith:fence_virsh):   Started rh83-dev01
  * Resource Group: dummy-group:
    * dummy-1   (ocf:heartbeat:Dummy):   FAILED rh83-dev01
  * Clone Set: prmPing-clone [prmPing]:
    * Started: [ rh83-dev01 rh83-dev02 rh83-dev03 ]

Node Attributes:
  * Node: rh83-dev01:
    * default_ping_set                  : 1         
  * Node: rh83-dev02:
    * default_ping_set                  : 1         
  * Node: rh83-dev03:
    * default_ping_set                  : 1         

Migration Summary:
  * Node: rh83-dev01:
    * dummy-1: migration-threshold=1 fail-count=1000000 last-failure='Mon Jun  7 09:14:19 2021'
  * Node: rh83-dev02:
    * fence3-virsh: migration-threshold=1 fail-count=1000000 last-failure='Mon Jun  7 09:14:29 2021'

Failed Resource Actions:
  * dummy-1_stop_0 on rh83-dev01 'error' (1): call=32, status='complete', exitreason='', last-rc-change='2021-06-07 09:14:19 +09:00', queued=0ms, exec=9ms
  * fence3-virsh_start_0 on rh83-dev02 'error' (1): call=27, status='complete', exitreason='', last-rc-change='2021-06-07 09:14:19 +09:00', queued=0ms, exec=10351ms

Failed Fencing Actions:
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:16:12 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:16:01 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:15:50 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:15:38 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:15:27 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:15:16 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:15:04 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:14:53 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:14:41 +09:00' 
  * reboot of rh83-dev01 failed: delegate=rh83-dev02, client=pacemaker-controld.4237, origin=rh83-dev01, completed='2021-06-07 09:14:30 +09:00' 
```

If the failing node is a DC node and the first fencing fails, escalation fencing will not be directed.
Therefore, failover will fail.

The problem occurs when fencing occurs targeting DC nodes in the absence of topology.

In this case, the fencing failure should be sent back to the DC node, but it will be notified by "Broadcast".
Therefore, the DC node does not perform fencing escalation.

This fix solves this problem.

Best Regards,
Hideo Yamauchi.

### CODE REVIEW

**Comment by kgaillot** [comment]
Hi Hideo,

Thanks for this fix! This is a good analysis, and the solution takes care of the problem you describe.

I am worried about the scenario where a malfunctioning node initiates its own fencing, which fails (as here), but the originator has completely stopped processing by the time the failure notification is sent. Before this fix, every node would get the notice of the failure. After this fix, I believe the operation would be "stuck" as pending.

It would be convenient if we could check the (currently ignored) return value of the send_cluster_message() call below your change, and resend the message as a broadcast if it failed, but send_cluster_message() only adds the message to a queue, so it doesn't tell us whether the originator received the message.

I'm thinking we could try to forward all self-fencing requests to another peer when possible, instead of only self-fencing requests with topology (see fenced_commands.c:handle_request() by "alternate_host"). I can't think of any serious problems with this approach.

Do you have any thoughts about this case?

**Comment by HideoYamauchi** [comment]
Hi Ken,

Thanks for your comment.
Please give me some time to think and see.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,

I thought about it for a moment.

> I am worried about the scenario where a malfunctioning node initiates its own fencing, which fails (as here), but the originator has completely stopped processing by the time the failure notification is sent. Before this fix, every node would get the notice of the failure. After this fix, I believe the operation would be "stuck" as pending.

This is also true of the successful case I have presented (unless it is a DC node failure).
Currently, even if it is not a DC node failure, it will be stuck until it is notified of the failure.
This is because the failure is not "broadcast" even if it is not a DC node failure.

Perhaps even if the send fails, the node that requested the fencing will time out and try to perform the next fencing. (Although confirmation is required)

> It would be convenient if we could check the (currently ignored) return value of the send_cluster_message() call below your change, and resend the message as a broadcast if it failed, but send_cluster_message() only adds the message to a queue, so it doesn't tell us whether the originator received the message.

I agree.
This should be synchronous communication or other message exchange mechanisms should be considered.

> 
> I'm thinking we could try to forward all self-fencing requests to another peer when possible, instead of only self-fencing requests with topology (see fenced_commands.c:handle_request() by "alternate_host"). I can't think of any serious problems with this approach.
> 

Certainly this may be the correct mechanism.
However, I suspect that the current fencing to be transferred and the fencing not to be transferred had some explicit intention of the developers in the past.
I think we need to check a little more to see if the changes will have any impact.

Best Regards,
Hideo Yamauchi.

**Comment by kgaillot** [comment]
> Hi Ken,
> 
> I thought about it for a moment.
> 
> > I am worried about the scenario where a malfunctioning node initiates its own fencing, which fails (as here), but the originator has completely stopped processing by the time the failure notification is sent. Before this fix, every node would get the notice of the failure. After this fix, I believe the operation would be "stuck" as pending.
> 
> This is also true of the successful case I have presented (unless it is a DC node failure).

In the case where rc==0, the result is always broadcast, so the other nodes will consider the operation complete.

process_remote_stonith_exec() calls remote_op_done(), which behaves differently for broadcast results and direct results. For broadcast results (only), it will notify local clients of the result and delete the operation.

> Currently, even if it is not a DC node failure, it will be stuck until it is notified of the failure.

By "stuck," I mean that a final notification is never broadcast, so the other nodes never consider the operation complete. It will show as pending forever.

If a notification is not broadcast but sent only to the originating node, the originating node does further processing if needed and eventually re-broadcasts the final notification. I suspect (but haven't confirmed) that if the originating node stops functioning before it receives the direct notification, then a broadcast will never happen, and the operation will be permanently pending. If that is correct, it's a problem regardless of whether it's a self-fencing request, but a node initiating a self-fencing request is more likely to be experiencing difficulties and stop functioning.

> This is because the failure is not "broadcast" even if it is not a DC node failure.

When an originator receives the direct notification, remote_op_done() calls stonith_bcast_result_to_peers() to resend the notification as a broadcast. When the other nodes receive that, they consider the operation complete.

> Perhaps even if the send fails, the node that requested the fencing will time out and try to perform the next fencing. (Although confirmation is required)

The originator is the node that requested the fencing. If it has stopped functioning before receiving the notification, no other node will time out the operation, as far as I'm aware. (The controller has a timeout on its end, but the operation will still be considered pending in the fencer.)

> > It would be convenient if we could check the (currently ignored) return value of the send_cluster_message() call below your change, and resend the message as a broadcast if it failed, but send_cluster_message() only adds the message to a queue, so it doesn't tell us whether the originator received the message.
> 
> I agree.
> This should be synchronous communication or other message exchange mechanisms should be considered.

The asynchronous queue is useful for serializing messages and for ensuring higher-priority work can be done in the main loop, but I could imagine having a callback after the data is sent, and deferring processing until then.

> > I'm thinking we could try to forward all self-fencing requests to another peer when possible, instead of only self-fencing requests with topology (see fenced_commands.c:handle_request() by "alternate_host"). I can't think of any serious problems with this approach.
> 
> Certainly this may be the correct mechanism.
> However, I suspect that the current fencing to be transferred and the fencing not to be transferred had some explicit intention of the developers in the past.

That's a good question. I suspect the relay was added for topology fencing simply because many topologies won't work otherwise (if the first device succeeds, the node may no longer be able to communicate with the cluster), rather than intentionally avoided for non-topology fencing.

> I think we need to check a little more to see if the changes will have any impact.
> 
> Best Regards,
> Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,

> > I thought about it for a moment.
> > > I am worried about the scenario where a malfunctioning node initiates its own fencing, which fails (as here), but the originator has completely stopped processing by the time the failure notification is sent. Before this fix, every node would get the notice of the failure. After this fix, I believe the operation would be "stuck" as pending.
> > 
> > 
> > This is also true of the successful case I have presented (unless it is a DC node failure).
> 
> In the case where rc==0, the result is always broadcast, so the other nodes will consider the operation complete.
Yes.

This was lacking in my explanation.

I said the same thing happens in a successful example in the following cases:
1) A resource failure occurs on node 1 other than the DC node.
2) Node 3 of the DC node requests fencing to node 2.
3) The fencing route of node 2 is blocked and fails.
4) Node 2 sends to the DC node, but again, it does not perform "broad cast" communication, so again, the same problem occurs if it cannot be sent.

> 
> process_remote_stonith_exec() calls remote_op_done(), which behaves differently for broadcast results and direct results. For broadcast results (only), it will notify local clients of the result and delete the operation.
> 
> > Currently, even if it is not a DC node failure, it will be stuck until it is notified of the failure.
> 
> By "stuck," I mean that a final notification is never broadcast, so the other nodes never consider the operation complete. It will show as pending forever.
> 
> If a notification is not broadcast but sent only to the originating node, the originating node does further processing if needed and eventually re-broadcasts the final notification. I suspect (but haven't confirmed) that if the originating node stops functioning before it receives the direct notification, then a broadcast will never happen, and the operation will be permanently pending. If that is correct, it's a problem regardless of whether it's a self-fencing request, but a node initiating a self-fencing request is more likely to be experiencing difficulties and stop functioning.

Let's stop the communication response in a pseudo manner and actually check what happens.
> 
> > This is because the failure is not "broadcast" even if it is not a DC node failure.
> 
> When an originator receives the direct notification, remote_op_done() calls stonith_bcast_result_to_peers() to resend the notification as a broadcast. When the other nodes receive that, they consider the operation complete.

Yes.

> 
> > Perhaps even if the send fails, the node that requested the fencing will time out and try to perform the next fencing. (Although confirmation is required)
> 
> The originator is the node that requested the fencing. If it has stopped functioning before receiving the notification, no other node will time out the operation, as far as I'm aware. (The controller has a timeout on its end, but the operation will still be considered pending in the fencer.)

Yes.

> 
> > > It would be convenient if we could check the (currently ignored) return value of the send_cluster_message() call below your change, and resend the message as a broadcast if it failed, but send_cluster_message() only adds the message to a queue, so it doesn't tell us whether the originator received the message.
> > 
> > 
> > I agree.
> > This should be synchronous communication or other message exchange mechanisms should be considered.
> 
> The asynchronous queue is useful for serializing messages and for ensuring higher-priority work can be done in the main loop, but I could imagine having a callback after the data is sent, and deferring processing until then.

I think so too.

> 
> > > I'm thinking we could try to forward all self-fencing requests to another peer when possible, instead of only self-fencing requests with topology (see fenced_commands.c:handle_request() by "alternate_host"). I can't think of any serious problems with this approach.
> > 
> > 
> > Certainly this may be the correct mechanism.
> > However, I suspect that the current fencing to be transferred and the fencing not to be transferred had some explicit intention of the developers in the past.
> 
> That's a good question. I suspect the relay was added for topology fencing simply because many topologies won't work otherwise (if the first device succeeds, the node may no longer be able to communicate with the cluster), rather than intentionally avoided for non-topology fencing.

Thank you for your good knowledge.
I certainly think so.

Best Regards,
Hideo Yamauchi.

**Comment by kgaillot** [comment]
Hi Hideo,

> > The asynchronous queue is useful for serializing messages and for ensuring higher-priority work can be done in the main loop, but I could imagine having a callback after the data is sent, and deferring processing until then.

Thinking about this some more, this might not be practical. Successfully sending the message (via cpg_mcast_joined()) does not guarantee the message was received; we can only be sure when we receive it back ourselves (via stonith_peer_ais_callback() for pacemaker-fenced).

Currently, there is a CPG message ID, but it is private to pcmk__cpg_msg_s in cpg.c. 

stonith_send_async_reply() would need to obtain that message ID somehow where it currently calls send_cluster_message() for a directed reply. A new function in libcrmcluster could maybe return the ID on success, but keeping the API backward-compatible might be difficult. The fencer would then store the message ID in a new table with whatever information is needed to process it later.

stonith_peer_ais_callback() would need a new libcrmcluster function to check whether the newly received message was sent *from* the local host, and get the message ID if so (before calling pcmk_message_common_cs(), which will ignore messages that aren't *to* the local host). It could then check the table to see if the message ID is present, and if so, do whatever processing is needed.

We might also need a new timeout in case the message is not received in a timely manner.

That all seems difficult, potentially adding significant overhead, so I think it is not worthwhile.

**Comment by wenningerk** [comment]
> Thinking about this some more, this might not be practical. Successfully sending the message (via cpg_mcast_joined()) does not guarantee the message was received; we can only be sure when we receive it back ourselves (via stonith_peer_ais_callback() for pacemaker-fenced).
> 

On top receiving a cpg-message back just tells us it was delivered to all members of the current membership - which might not include the intended receiver - right?

**Comment by HideoYamauchi** [comment]
Hi Ken,
Hi Klaus,

I'll see if there's a solution in a little more detail later next week.
I'm taking a lot of time to do other work, so please take a moment.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,
Hi Klaus,

I feel bad.
I don't think I'll have time this week either.
Let me discuss it again next week.

Let me check the problem again next week as my perception may be wrong.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,
Hi Klaus,

Next week, I'll finally be able to return to this PR discussion.
I was a little too far from the topic.
I will review this issue once.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,
Hi Klaus,

I've been away from this discussion for a while, so let me check it once.

---
I did this PR because it was an N-node configuration without topology and fencing did not escalate.
Certainly, as you have pointed out, problems arise when the fencing organizers are solidified.

As I wrote in the previous comment, even with toplogy, you may only send a failure message to the fencing organizer.
Therefore, even if there is topology, the problem pointed out will occur.

I think our discussion now is not only about this PR issue, but also about ensuring that the message of failure is sent to the fencing organizer, with or without topology.

---

Is my understanding wrong?


Best Regards,
Hideo Yamauchi.

**Comment by kgaillot** [comment]
> Hi Ken,
> Hi Klaus,
> 
> I've been away from this discussion for a while, so let me check it once.
> 
> I did this PR because it was an N-node configuration without topology and fencing did not escalate.
> Certainly, as you have pointed out, problems arise when the fencing organizers are solidified.
> 
> As I wrote in the previous comment, even with toplogy, you may only send a failure message to the fencing organizer.
> Therefore, even if there is topology, the problem pointed out will occur.
> 
> I think our discussion now is not only about this PR issue, but also about ensuring that the message of failure is sent to the fencing organizer, with or without topology.
> 
> Is my understanding wrong?
> 
> Best Regards,
> Hideo Yamauchi.

Hi Hideo,

The fencer's control flow is hard to follow, but I think we have a general problem, and a specific case of it.

The general problem is that if a stonith result is *not* broadcast, but sent directly to the originator, and the originator is not able to receive or process that result for any reason, then the request will be forever "stuck" as pending in the fencing history. This is true even if the controller times out the operation on its side and submits a new request.

The specific case is that with this PR, failure notifications for self-initiated fencing will be sent only to the initiating node. In this situation, something is clearly wrong with the initiating node, otherwise it wouldn't need to be fenced. This means the general problem is more likely to happen, because the initiating node is more likely to stop functioning after requesting fencing.

One way to mitigate the specific case would be to forward all self-fencing requests (even those without topology) to another node if one is available. This would greatly reduce the occurrence of self-initiated fencing in general, reducing the opportunity for the problem to arise.

Avoiding the general case would be more difficult. One possibility would be for the node that is chosen to execute fencing to set a timer, and if it does not receive a broadcast within that time, it would broadcast the result itself. Another possibility would be for all nodes to set such a timer, and report a failed result locally if a broadcast is not received in time. That might be possible only for non-topology requests, though, and choosing the timeout value might be a challenge.

**Comment by HideoYamauchi** [comment]
Hi Ken,

Thanks for your comment.

> > Hi Ken,
> > Hi Klaus,
> > I've been away from this discussion for a while, so let me check it once.
> > I did this PR because it was an N-node configuration without topology and fencing did not escalate.
> > Certainly, as you have pointed out, problems arise when the fencing organizers are solidified.
> > As I wrote in the previous comment, even with toplogy, you may only send a failure message to the fencing organizer.
> > Therefore, even if there is topology, the problem pointed out will occur.
> > I think our discussion now is not only about this PR issue, but also about ensuring that the message of failure is sent to the fencing organizer, with or without topology.
> > Is my understanding wrong?
> > Best Regards,
> > Hideo Yamauchi.
> 
> Hi Hideo,
> 
> The fencer's control flow is hard to follow, but I think we have a general problem, and a specific case of it.
> 
> The general problem is that if a stonith result is _not_ broadcast, but sent directly to the originator, and the originator is not able to receive or process that result for any reason, then the request will be forever "stuck" as pending in the fencing history. This is true even if the controller times out the operation on its side and submits a new request.
> 

Just I've seen another issue behave like this.


> The specific case is that with this PR, failure notifications for self-initiated fencing will be sent only to the initiating node. In this situation, something is clearly wrong with the initiating node, otherwise it wouldn't need to be fenced. This means the general problem is more likely to happen, because the initiating node is more likely to stop functioning after requesting fencing.
> 
> One way to mitigate the specific case would be to forward all self-fencing requests (even those without topology) to another node if one is available. This would greatly reduce the occurrence of self-initiated fencing in general, reducing the opportunity for the problem to arise.
> 
> Avoiding the general case would be more difficult. One possibility would be for the node that is chosen to execute fencing to set a timer, and if it does not receive a broadcast within that time, it would broadcast the result itself. Another possibility would be for all nodes to set such a timer, and report a failed result locally if a broadcast is not received in time. That might be possible only for non-topology requests, though, and choosing the timeout value might be a challenge.

I understood.
I'll think a little more if there is a better fix.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,
Hi Klaus,

I thought about it, but there seems to be no other good way.

>One possibility would be for the node that is chosen to execute fencing to set a timer, and if it does not receive a broadcast within that time, it would broadcast the result itself. 

I would like to try this method once.
After that, let me continue the discussion with the new PR.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,
Hi Klaus,

I'm late.
I will go into creating this fix.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Ken,
Hi Klaus,

I will publish a WIP PR for discussion next week.
I have a few concerns, so please let me know what you think.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi All,

I was late for another job to get to work on this issue.
I will review the latest fenced source and resume fixing this issue.

Best Regards,
Hideo Yamauchi.
